src/polyfills/deno-wasi.ts

```diff
@@ -6,25 +6,34 @@ import { closeSync } from 'node:fs';
 
 async function createDevNullFDs() {
   const [stdin, stdout] = await Promise.all([open(devNull, 'r'), open(devNull, 'w')]);
-
+  let needsClose = true;
   const fr = new globalThis.FinalizationRegistry((held: number) => {
     try {
-      closeSync(held);
+      if (needsClose) closeSync(held);
     } catch {
       // The fd may already be closed.
     }
   });
   fr.register(stdin, stdin.fd);
   fr.register(stdout, stdout.fd);
 
-  return [stdin.fd, stdout.fd, stdout.fd];
+  return {
+    async close() {
+      needsClose = false;
+      await Promise.all([stdin.close(), stdout.close()]).catch(() => {});
+    },
+    fds: [stdin.fd, stdout.fd, stdout.fd],
```

Comment:
Deno, unlike Node, runs its "leaked file descriptor" check _before_ the `FinalizationRegistry` has a chance to fire; so we add the explicit close here.